graql/reasoner/query/DisjunctiveQuery.java

```diff
@@ -0,0 +1,304 @@
+    @Override
+    public Set<String> validateOntologically(Label ruleLabel) {
+        return getClauses().stream().flatMap(q -> q.validateOntologically(ruleLabel).stream()).collect(Collectors.toSet());
+    }
+
+    public HashMap<Variable, Concept> filterBindingVars(Map<Variable, Concept> map) {
+
+        HashMap<Variable, Concept> bindingVarsMap = new HashMap<>();
+
+        bindingVars.forEach(b -> {
+            if (map.get(b) != null) {
+                bindingVarsMap.put(b, map.get(b));
+            }
+        });
+
+        return bindingVarsMap;
+    }
+
+    @Override
+    public ResolvableQuery withSubstitution(ConceptMap sub) {
+        return new DisjunctiveQuery(
+                getClauses().stream().map(q -> q.withSubstitution(sub)).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    public Stream<ConceptMap> resolve(Set<ReasonerAtomicQuery> subGoals, boolean infer) {
+
+        boolean doNotResolve = !infer || getAtoms().isEmpty() || (isPositive() && !isRuleResolvable());
+        if (doNotResolve) {
+            Stream<Stream<ConceptMap>> answerStreams = clauses.stream().map(clause ->
+                    traversalExecutor.traverse(clause.getPattern()).map(ans -> {
+                        ConceptMap clauseAns = new ConceptMap(ans.map(), new LookupExplanation(), clause.getPattern(ans.map()));
+                        HashMap<Variable, Concept> bindingVarsSub = filterBindingVars(ans.map());
+                        return new ConceptMap(bindingVarsSub, new DisjunctiveExplanation(clauseAns), getPattern(bindingVarsSub));
+                    }));
+            LazyMergingStream<ConceptMap> mergedStreams = new LazyMergingStream<>(answerStreams);
+            return mergedStreams.flatStream();
+        } else {
+            return new ResolutionIterator(this, subGoals, context().queryCache()).hasStream();
+        }
+    }
+
+    @Override
+    ResolvableQuery inferTypes() {
+        return new DisjunctiveQuery(
+                getClauses().stream().map(CompositeQuery::inferTypes).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    ResolvableQuery constantValuePredicateQuery() {
+        return new DisjunctiveQuery(
+                getClauses().stream().map(CompositeQuery::constantValuePredicateQuery).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    ResolvableQuery copy() {
+        return new DisjunctiveQuery(
+                new HashSet<>(getClauses()),
+                getBindingVars(),
+                traversalExecutor,
+                context());
+    }
+
+    @Override
+    public boolean isAtomic() {
+        // TODO unclear of the meaning of atomicity in this case
```

Comment:
@kasper-piskorski what do you think the behaviour of `isAtomic()` should be for a Disjunctive Query?

Reply:
atomic queries are either `ReasonerAtomicQuery` or conjunctions (`ReasonerQueryImpl`) with a single selectable atom. Since `DisjunctiveQuery` contains always >1 conjunctions it is never atomic. So either return false or throw - disjunctions being expected to be atomic might suggests problems in the code and a behaviour to be investigated.